docs/Debugging.md

```diff
@@ -0,0 +1,85 @@
+
+The instructions below assume that Open Enclave was installed in the default directory `/opt/openenclave`.
+If you used a different directory, run `. <OE_PREFIX>/share/openenclave/openenclaverc` first.
+
+### Build mode
+
+Debugging enclave code (SGX-LKL as well as the application code) is only possible in debug and non-release builds of SGX-LKL.
+
+### Encrypted disks
+
+TODO add note on how to handle situations when disk keys are retrieved from a remote service  (either re-create disks unencrypted, re-create disks encrypted but with debug-enabled key release policy (closest to production), or hard-code key in app config)
+
+### Application type
+
+Debugging of non-native applications like Python, Java or .NET applications is currently not possible.
+
+### Docker deployment container
+
+The instructions below assume that disk images and configuration are available outside of a Docker deployment container. Debugging of such deployment containers is not supported currently as they do not contain the necessary debugging tools.
+
+## Option A: Debugging with an SGX-LKL installation
+
+The instructions below assume that SGX-LKL was installed in `/opt/sgx-lkl`.
+
+To debug an application, prefix the regular `sgx-lkl-run-oe` command with `sgx-lkl-gdb --args`, for example:
+```
+/opt/sgx-lkl/bin/sgx-lkl-gdb --args /opt/sgx-lkl/bin/sgx-lkl-run-oe --hw-debug root_disk.img /app
+```
+See `sgx-lkl-run-oe --help` for how to run using host and app config files.
+
+## Option B: Debugging with an SGX-LKL source build
+
+This assumes you want to debug SGX-LKL directly from the source tree without installing it.
+
+To debug an application, prefix the regular `sgx-lkl-run-oe` command with `sgx-lkl-gdb --args`, for example:
+```
+tools/gdb/sgx-lkl-gdb --args build/sgx-lkl-run-oe --hw-debug root_disk.img /app
+```
+See `sgx-lkl-run-oe --help` for how to run using host and app config files.
+
+## Debugging advice
+
+### Visual Studio Code
+
+Visual Studio Code allows to use a custom `gdb` command for debugging.
+To use `sgx-lkl-gdb`, add `"miDebuggerPath": "/opt/sgx-lkl/bin/sgx-lkl-gdb"` (change path if necessary) in your `.vscode/launch.json` file.
+Tip: Use the "Visual Studio Code Remote - SSH" extension to debug on a remote VM with SGX support.
+
+### Run mode
+
+When using `--sw-debug` then SGX hardware is simulated. However, not all aspects are simulated, for example, certain CPU instructions are invalid within SGX, but would execute fine in simulation mode.
+If possible, debugging should be done in hardware mode to be as close to a production environment as possible.
+
+### Ignore SIGILL
+
+In some cases `sgx-lkl-gdb` still incorrectly stops at illegal instructions that are emulated inside the enclave (see e.g. https://github.com/lsds/sgx-lkl/issues/167). To avoid that, run `handle SIGILL pass nostop noprint` as first command in the active `gdb` session.
```

Comment:
```suggestion
In some cases `sgx-lkl-gdb` stops at illegal instructions that are emulated inside the enclave (see e.g. https://github.com/lsds/sgx-lkl/issues/167). To avoid that, run `handle SIGILL pass nostop noprint` as the first command in the active `gdb` session.
```